pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java

```diff
@@ -0,0 +1,153 @@
+        super();
+
+        this.enabled = conf.isBatchingEnabled();
+        this.maxDelayMillis = conf.getBatchingMaxDelayMillis();
+        this.maxOperations = conf.getBatchingMaxOperations();
+        this.maxSize = conf.getBatchingMaxSizeKb() * 1_024;
+
+        readOps = new MpscUnboundedArrayQueue<>(10_000);
+        writeOps = new MpscUnboundedArrayQueue<>(10_000);
+        if (enabled) {
+            scheduledTask = executor.scheduleAtFixedRate(this::flush, 5, 5, TimeUnit.MILLISECONDS);
+        } else {
+            scheduledTask = null;
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (scheduledTask != null) {
+            scheduledTask.cancel(true);
+        }
+        super.close();
+    }
+
+    private void flush() {
+        while (!readOps.isEmpty()) {
+            List<MetadataOp> ops = new ArrayList<>();
+            readOps.drain(ops::add, maxOperations);
+            batchOperation(ops);
+        }
+
+        while (!writeOps.isEmpty()) {
+            int batchSize = 0;
+
+            List<MetadataOp> ops = new ArrayList<>();
+            for (int i = 0; i < maxOperations; i++) {
+                MetadataOp op = writeOps.peek();
+                if (op == null) {
+                    break;
+                }
+
+                if (i > 0 && (batchSize + op.size()) > maxSize) {
+                    // We have already reached the max size, so flush the current batch
+                    break;
+                }
+
+                batchSize += op.size();
+                ops.add(writeOps.poll());
+            }
+            batchOperation(ops);
+        }
+
+        flushInProgress.set(false);
+    }
+
+    @Override
+    public final CompletableFuture<Optional<GetResult>> storeGet(String path) {
+        OpGet op = new OpGet(path);
+        enqueue(readOps, op);
+        return op.getFuture();
+    }
+
+    @Override
+    protected final CompletableFuture<List<String>> getChildrenFromStore(String path) {
+        OpGetChildren op = new OpGetChildren(path);
+        enqueue(readOps, op);
+        return op.getFuture();
+    }
+
+    @Override
+    protected final CompletableFuture<Void> storeDelete(String path, Optional<Long> expectedVersion) {
+        OpDelete op = new OpDelete(path, expectedVersion);
+        enqueue(writeOps, op);
+        return op.getFuture();
+    }
+
+    @Override
+    protected final CompletableFuture<Stat> storePut(String path, byte[] data, Optional<Long> optExpectedVersion,
+                                               EnumSet<CreateOption> options) {
+        OpPut op = new OpPut(path, data, optExpectedVersion, options);
+        enqueue(writeOps, op);
+        return op.getFuture();
+    }
+
+    private void enqueue(MessagePassingQueue queue, MetadataOp op) {
+        if (enabled) {
+            if (!queue.offer(op)) {
+                op.getFuture().completeExceptionally(new IllegalStateException("metadata queue is full"));
```

Comment:
Os it better to fail with an error of the Metadata Store Exceptions family?
This way caller code won't see surprises with RuntimeException